sql/table.cc

```diff
@@ -751,7 +751,9 @@ void setup_key_part_field(TABLE_SHARE *share, handler *handler_file,
   } else if (part_of_key_not_extended) {
     field->part_of_prefixkey.set_bit(key_n);
   }
-  if ((handler_file->index_flags(key_n, key_part_n, 0) & HA_KEYREAD_ONLY) &&
+  /* call index_flags() only after all key parts were processed */
+  if ((key_part_n == keyinfo->actual_key_parts - 1) &&
```

Comment:
I'm seeing some test failures that look like the wrong result based on this change.

```
CURRENT_TEST: rocksdb.type_varchar
--- /mnt/btrfs/trunk-git-mysql-111611707-1596472507/mysql-test/suite/rocksdb/r/type_varchar.result	2020-08-03 19:35:41.696722657 +0300
+++ /mnt/btrfs/trunk-git-mysql-111611707-1596472507/_build-8.0-Release/mysql-test/var/16/log/type_varchar.reject	2020-08-03 20:20:57.945493924 +0300
@@ -444,28 +444,24 @@
 explain
 select col1, hex(col1) from t1 force index(col1);
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	t1	NULL	index	NULL	col1	131	NULL	#	#	Using index
+1	SIMPLE	t1	NULL	ALL	NULL	NULL	NULL	NULL	#	#	NULL
```

I thought there is a comment in ha_rocksdb::index_flags on the function being called by the SQL layer before the tables are necessarily fully setup. However, it seems like changing the code here leads to other failures.

Do you have the test case for 8.0 Sk always showing using index when it should not be? Is there a better fix that can be made inside of ha_rocksdb::index_flags() instead?

Reply:
To check that 8.0 Sk always shows `using index` please just import: 
```
	new file:   mysql-test/suite/rocksdb/r/type_char_secondary_index_only_scan.result
	new file:   mysql-test/suite/rocksdb/t/type_char_secondary_index_only_scan.test
```
from https://github.com/facebook/mysql-5.6/commit/fb4549660c

Reply:
I will work on moving a fix from `sql/table.cc` to `ha_rocksdb::index_flags`.
BTW, do you run MTR tests with ` --character-set-server` and `--collation-server`?
I see a lot of issues in `rocksdb.type_*` tests:
```
CURRENT_TEST: rocksdb.type_varchar
--- /data/mysql-server/fb-8.0.17/mysql-test/suite/rocksdb/r/type_varchar.result	2020-07-29 16:22:29.948380928 +0300
+++ /data/mysql-server/fb-8.0.17-deb-clang9/mysql-test/var/log/type_varchar.reject	2020-08-10 15:26:19.743306454 +0300
@@ -16,16 +16,22 @@
 v65000 VARCHAR(65000) ,
 PRIMARY KEY (v64)
 ) ENGINE=rocksdb;
+Warnings:
+Note	1246	Converting column 'v65000' from VARCHAR to TEXT
 SHOW COLUMNS IN t1;
 Field	Type	Null	Key	Default	Extra
 v0	varchar(0)	YES		NULL	
 v1	varchar(1)	YES		NULL	
 v64	varchar(64)	NO	PRI	NULL	
-v65000	varchar(65000)	YES		NULL	
-CREATE TABLE t2 (v VARCHAR(65532), PRIMARY KEY (v(255))) ENGINE=rocksdb;
+v65000	mediumtext	YES		NULL	
+CREATE TABLE t2 (v VARCHAR(65532), PRIMARY KEY (v(255))) ENGINE=rocksdb CHARSET utf8 COLLATE utf8_general_ci;
+Warnings:
+Warning	3719	'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous.
+Warning	3778	'utf8_general_ci' is a collation of the deprecated character set UTF8MB3. Please consider using UTF8MB4 with an appropriate collation instead.
+Note	1246	Converting column 'v' from VARCHAR to TEXT
 SHOW COLUMNS IN t2;
 Field	Type	Null	Key	Default	Extra
-v	varchar(65532)	NO	PRI	NULL	
+v	mediumtext	NO	PRI	NULL	
```

Reply:
It's not clear to me if the following query plan using filesort is problematic for `select id from t3 order by vch;`, vs the current solution (use index for ordering and fetch the primary key on each row when needed).

```
create table t3 (id int NOT NULL AUTO_INCREMENT PRIMARY KEY, vch VARCHAR(64), ch CHAR(64), KEY vch_idx(vch), key ch_idx(ch)) ENGINE=rocksdb CHARSET utf8mb4 collate=utf8mb4_croatian_ci;

mysql> mysql> explain select id from t3 order by vch;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+----------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra          |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+----------------+
|  1 | SIMPLE      | t3    | NULL       | ALL  | NULL          | NULL | NULL    | NULL |   15 |   100.00 | Using filesort |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+----------------+
1 row in set, 1 warning (0.00 sec)
```

However, bd907766e basically addresses the issue in the general case and provides the fix needed so that Using index is accurate. Given this, is a targeted fix in ha_rocksdb.cc still needed?

Since this change in sql/table.cc is causing other potential issues, I'm considering dropping it.

And yes, we do have different settings for the rocksdb suite to maintain the backward compatibility with 5.6 test results. 3976b20b5c introduced the settings in mysql-test/suite/rocksdb/my.cnf.